src/NUnitEngine/nunit.engine/Communication/Transports/Tcp/TestAgentTcpProxy.cs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) Charlie Poole, Rob Prouse and Contributors. MIT License - see LICENSE.txt
+
+using System;
+using System.Net.Sockets;
+using NUnit.Engine.Communication.Messages;
+using NUnit.Engine.Communication.Protocols;
+
+namespace NUnit.Engine.Communication.Transports.Tcp
+{
+    /// <summary>
+    /// TestAgentTcpProxy wraps a RemoteTestAgent so that certain
+    /// of its properties may be accessed directly.
+    /// </summary>
+    internal class TestAgentTcpProxy : ITestAgent, ITestEngineRunner
+    {
+        private Socket _socket;
+        private BinarySerializationProtocol _wireProtocol = new BinarySerializationProtocol();
+
+        public TestAgentTcpProxy(Socket socket, Guid id)
+        {
+           _socket = socket;
+            Id = id;
+        }
+
+        public Guid Id { get; }
+
+        public ITestEngineRunner CreateRunner(TestPackage package)
+        {
+            SendCommandMessage("CreateRunner", package);
+
+            // Agent also functions as the runner
+            return this;
+        }
+
+        public bool Start()
+        {
+            SendCommandMessage("Start");
+            return CommandResult<bool>();
+        }
+
+        public void Stop()
+        {
+            SendCommandMessage("Stop");
+        }
+
+        public TestEngineResult Load()
+        {
+            SendCommandMessage("Load");
+            return CommandResult<TestEngineResult>();
+        }
+
+        public void Unload()
+        {
+            SendCommandMessage("Unload");
+        }
+
+        public TestEngineResult Reload()
+        {
+            SendCommandMessage("Reload");
+            return CommandResult<TestEngineResult>();
+        }
+
+        public int CountTestCases(TestFilter filter)
+        {
+            SendCommandMessage("CountTestCases", filter);
+            return CommandResult<int>();
+        }
+
+        public TestEngineResult Run(ITestEventListener listener, TestFilter filter)
+        {
+            SendCommandMessage("Run", filter);
+
+            return TestRunResult(listener);
+        }
+
+        public AsyncTestEngineResult RunAsync(ITestEventListener listener, TestFilter filter)
+        {
+            SendCommandMessage("RunAsync", filter);
+            // TODO: Should we get the async result from the agent or just use our own?
+            return CommandResult<AsyncTestEngineResult>();
+            //return new AsyncTestEngineResult();
+        }
+
+        public void StopRun(bool force)
+        {
+            SendCommandMessage("StopRun", force);
+        }
+
+        public TestEngineResult Explore(TestFilter filter)
+        {
+            SendCommandMessage("Explore", filter);
+            return CommandResult<TestEngineResult>();
+        }
+
+        public void Dispose()
+        {
+            throw new NotImplementedException();
```

Comment:
Should this be empty, rather than not implemented exception? Or is there something here that does need disposing? I seem to remember we had a need to dispose runners at some point...can't remember why it was though however.

Reply:
I'll check this one.

Reply:
I put in the fix you suggested, 'cause it works! :smile: 

However, I went down a bit of a rabbit hole before settling on this and I think we ought to revisit it after this is merged, as a part of reviewing out API interfaces.

Specifically, `TestAgentTcpProxy` needs `Dispose()` because it implements `ITestEngineRunner`, which in turn derives from `IDisposable`. I spent a lot of time pondering this and I now think the latter derivation is a mistake.

When a _class_ implements `IDisposable`, it means that the class holds resources that need to be released, either directly or indirectly. But when an _interface_ derives from `IDisposable`, it says that _every_ class implementing it will have resources that need to be released. After a lot of thought, I believe that's almost never true.

This is a great example. The proxy implements `ITestAgentRunner` in order to return itself _as a runner_. But of course it only _looks like_ a runner, which is exactly what a proxy does. it has nothing to dispose.

All that said, for version 3.x we can only do what you have suggested. For version 4.0, I think we should look at cases where any interface derives from `IDisposable` very critically and consider removing it in favor of having each class that needs to implement it. There will be tradeoffs in each case, of course.

I could create an issue for this @ChrisMaddock but I'm not sure if it should be a broad one (Review our API) or a narrower one as described here. What do you think?

Reply:
I think we should tackle this for 4.X. Would you be ok to create this issue? I think you have a clearer understanding of the core issue here than I do right now.